### PR TITLE
Small optimization for adam

### PR DESCRIPTION
**Pull request by jma127**
Apply weight decay for Adam in-place instead of via copy.

Synced offline with @soumith , who mentioned that it should be OK. This is also consistent with other optimizers, e.g. https://github.com/pytorch/pytorch/blob/eee01731a5d33d5be58d875711bd2577e38dbddf/torch/optim/sgd.py#L93

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Comment by SsnL** [comment]
well... I would certainly expect .grad to not change after optimizer step.

**Comment by jma127** [comment]
Hmm, then the SGD implementation should be fixed to satisfy that invariant.

I'll leave it to you guys to determine whether or not this is a necessary invariant -- feel free to revert as you see fit.
